### PR TITLE
Workday employee daily history model + monthly summary model + updates to employee surrogate key

### PR DESCRIPTION
**Pull request by fivetran-avinash**
## PR Overview
**This PR will address the following Issue/Feature:** N/A

**This PR will result in the following new package version:** 2.0

Adding surrogate keys to specific models, as well as history functionality which will rapidly change the nature of the package. 

**Please provide the finalized CHANGELOG entry which details the relevant changes included in this PR:**

# dbt_workday v0.2.0
## 🚨 Breaking Changes 🚨
- Created a surrogate key `employee_id` in `workday__employee_overview` that combines `worker_id`, `position_id`, and `position_start_date`. This accounts for the edge cases where:
  - A worker can hold multiple positions concurrently.
  - A position being held by multiple workers concurrently.
  - A worker being rehired for the same position. 
- Using this surrogate key as our grain will hopefully provide uniqueness for the majority of Workday HCM customer cases. 

## 🚀 Feature Updates 🚀 
- We have added staging history mode models in the [`models/history`](https://github.com/fivetran/dbt_workday/tree/main/models/staging/history) folder [to support Fivetran's history mode feature](https://fivetran.com/docs/core-concepts/sync-modes/history-mode). 

This will allow customers to utilize the Fivetran history mode feature, which records every version of each record in the source table from the moment this mode is activated in the equivalent tables. 

These staging models include:

  - `stg_workday__personal_information_history`: Containing historical records of a worker's personal information.
  - `stg_workday__worker_history`: Containing historical records of a worker's history.
  - `stg_workday__worker_position_history`: Containing historical records of a worker's position history.
  - `stg_workday__worker_position_organization_history`: Containing historical records of a worker's position history.

- We have then utilized the `workday__employee_daily_history` model in the [`models/workday_history`](https://github.com/fivetran/dbt_workday/tree/main/models/workday_history) folder [based off of Fivetran's history mode feature](https://fivetran.com/docs/core-concepts/sync-modes/history-mode), pulling from Workday HCM source models you can view in the [`models/staging/workday_history`](https://github.com/fivetran/dbt_workday/tree/main/models/staging/workday_history) folder.

- These models are disabled by default due to their size, so you will need to set the below variable configurations for each of the individual models you want to utilize in your `dbt_project.yml`.

```yml
vars:
   employee_history_enabled: true  ##Ex: employee_history_enabled: true    
```

- We have also added the `workday__monthly_summary` model in the [`models/workday_history`](https://github.com/fivetran/dbt_workday/tree/main/models/workday_history) folder. This table aggregates high-level monthly metrics to track changes over time to overall employee data for a customer. 

- We have chosen not to implement incremental logic in the history models due to the future-facing updating of Workday HCM transactions beyond current daily updates. [See the DECISIONLOG](https://github.com/fivetran/dbt_workday/blob/main/DECISIONLOG.md) for more details.

- We support the option to pull from both your Workday HCM and History Mode connectors simultaneously from their specific database/schemas.  We also support pulling from just your History Mode connector on its own and bypassing the standard connector on its own. [See more detailed instructions in the README](https://github.com/fivetran/dbt_workday/blob/main/README.md#configuring-your-workday-history-mode-database-and-schema-variables). 

- Workday HCM History Mode models can contain a multitude of rows if you bring in all historical data, so we've introduced the flexibility to set first date filters to bring in only the historical data you need. [More details can be found in the README](https://github.com/fivetran/dbt_workday/blob/main/README.md#filter-your-workday-hcm-history-mode-models).

### Basic Validation
Please acknowledge that you have successfully performed the following commands locally:
- [x] dbt run –full-refresh && dbt test
- [NA] dbt run (if incremental models are present) && dbt test

Before marking this PR as "ready for review" the following have been applied:
- [NA] The appropriate issue has been linked, tagged, and properly assigned.
- [x] All necessary documentation and version upgrades have been applied.
    
- [x] docs were regenerated (unless this PR does not include any code or yml updates).
- [x] BuildKite integration tests are passing.
- [x] Detailed validation steps have been provided below.

### Detailed Validation
Please share any and all of your validation steps: Models currently testing and passing all warehouses with accurate daily data.

### If you had to summarize this PR in an emoji, which would it be?
 
⛑️

### CODE REVIEW

File: CHANGELOG.md
Comment (fivetran-joemarkiewicz) [inline CHANGELOG.md:3]: I think it is relevant to include this in the release notes, but what in particular makes this a breaking change?

The release should definitely be a minor `0.X.0` release, but not sure if this is a breaking change. What are your thoughts?

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-avinash thanks for all your hard work on this feature update! This is looking good. I do have some comments, suggestions, and questions to be addressed below. Two major items to callout include:
- There seems to be a lot of documentation around using two different schemas when using history mode. Since the history tables are in the same schema (and are the same sources we are using for the non history models), we don't need to add any documentation around using two different schemas for history and non history mode.
- We should try and adjust the datespines to be a bit more dynamic to reflect the actual data. The variable are going to come in handy, but I think we could make it more dynamic.

Thanks again, and let me know if you have any questions when going through the review.

File: .buildkite/scripts/run_models.sh
Comment (fivetran-avinash) [reply .buildkite/scripts/run_models.sh:22]: We took the incremental strategy out though? I can add it back in if needed.

File: integration_tests/dbt_project.yml
Comment (fivetran-avinash) [reply integration_tests/dbt_project.yml:36]: Good call-out. I'd set this to true for regenerating docs since otherwise the new models will not appear. But I will  use this as a reminder to set to false before merging.

File: models/staging/workday_history/stg_workday__worker_position_organization_history.sql
Comment (fivetran-avinash) [reply models/staging/workday_history/stg_workday__worker_position_organization_history.sql:1]: There is not. Flipping to false.

File: models/workday_history/workday__monthly_summary.sql
Comment (fivetran-avinash) [reply models/workday_history/workday__monthly_summary.sql:9]: Good call. This was mostly for development/validation purposes (since otherwise everything was very jumbled), but isn't necessary in the final model.

File: models/workday_history/workday__monthly_summary.sql
Comment (fivetran-avinash) [reply models/workday_history/workday__monthly_summary.sql:112]: Good call. This was mostly for development/validation purposes (since otherwise everything was very jumbled), but isn't necessary in the final model.

File: models/intermediate/workday_history/int_workday__employee_history.sql
Comment (fivetran-avinash) [reply models/intermediate/workday_history/int_workday__employee_history.sql:91]: The behavior Tommy explained was that a worker can hold multiple positions and (rarer) a position can hold multiple workers.

And additionally if a worker is rehired for the same position, they could fill in the same `worker_id` and `position_id`. 

So to achieve true uniqueness for a Workday employee, it is important to bring in the `worker_id`, `position_id` and `position_start_date`.

File: CHANGELOG.md
Comment (fivetran-avinash) [reply CHANGELOG.md:3]: Yeah, I went back and forth on this, as it isn't technically a standard breaking change. But we are updating the primary key. 

Because I was adding a new primary key, changing the grain of the model from `worker_id` to `employee_id`, it felt like a significant enough change to warrant as breaking? Or is there something else we should call it out as like 

🔑 New Primary Key 🔑 ?

File: dbt_project.yml
Comment (fivetran-avinash) [reply dbt_project.yml:11]: Think this was just for development, I'll flip this back.

File: README.md
Comment (fivetran-avinash) [reply README.md:37]: Good question, let's discuss this. I'm concerned about setting this variable to true to Quickstart at this time, given Tommy's concerns about the current Workday History data.

File: DECISIONLOG.md
Comment (fivetran-avinash) [reply DECISIONLOG.md:8]: Agreed, this is an artifact of the incremental documentation that was ported from Salesforce, but was not removed.

File: README.md
Comment (fivetran-avinash) [reply README.md:102]: Apologies, not sure what this is referring to in the above text 🙂

File: README.md
Comment (fivetran-avinash) [reply README.md:109]: Removing.

File: README.md
Comment (fivetran-avinash) [reply README.md:115]: Ack, yes. Removing.

File: CHANGELOG.md
Comment (fivetran-avinash) [reply CHANGELOG.md:41]: Removed

File: README.md
Comment (fivetran-avinash) [reply README.md:37]: After our discussion, I'll remove this for now and we can reconvene on this once Elan weighs in.

File: README.md
Comment (fivetran-avinash) [reply README.md:44]: Added.

File: README.md
Comment (fivetran-avinash) [reply README.md:98]: Updated.

File: README.md
Comment (fivetran-avinash) [reply README.md:102]: Did you mean this step is not needed?

File: dbt_project.yml
Comment (fivetran-avinash) [reply dbt_project.yml:16]: I renamed to `stg_workday_history`, let me know if there's a better name that makes sense!

And agreed again! Just another artifact of development. Adding yml materialization for the staging folder as well.

File: models/workday_history/workday__employee_daily_history.sql
Comment (fivetran-avinash) [reply models/workday_history/workday__employee_daily_history.sql:50]: Yes, the partition is based on `_fivetran_date` and ordered on `_fivetran_start` desc [in the subsequent CTE](https://github.com/fivetran/dbt_workday/blob/release/workday-history-upgrades/models/workday_history/workday__employee_daily_history.sql#L36-L44). So if there are multiple `_fivetran_start` values  for an `employee_id` on the same day, we can capture them and exclude the earlier records.

File: models/workday_history/workday__employee_daily_history.sql
Comment (fivetran-avinash) [reply models/workday_history/workday__employee_daily_history.sql:19]: Reconfigured the logic to go for the min(`_fivetran_start`) value and refresh as the second option! Let me know your thoughts.

File: models/workday_history/workday__employee_daily_history.sql
Comment (fivetran-avinash) [reply models/workday_history/workday__employee_daily_history.sql:6]: Good point! I reconfigured the logic to hopefully account for this. Let me know your thoughts.

File: models/workday_history/workday__worker_position_org_daily_history.sql
Comment (fivetran-avinash) [reply models/workday_history/workday__worker_position_org_daily_history.sql:3]: Agreed, modified to match the logic I applied in `employee_daily_history`

Comment (fivetran-avinash) [reply models/workday_history/workday__worker_position_org_daily_history.sql:3]: Agreed, modified to match the logic I applied in `employee_daily_history`.

File: models/workday_history/workday__monthly_summary.sql
Comment (fivetran-avinash) [reply models/workday_history/workday__monthly_summary.sql:6]: Casted.

File: models/staging/workday_history/stg_workday__personal_information_history.sql
Comment (fivetran-avinash) [reply models/staging/workday_history/stg_workday__personal_information_history.sql]: This should hopefully be accounted for now.

File: models/docs.md
Comment (fivetran-avinash) [reply models/docs.md:9]: Did a second pass and hopefully all fields should be documented now.

**Review by fivetran-avinash — COMMENTED** [review]
@fivetran-joemarkiewicz This is ready for re-review.

File: .buildkite/scripts/run_models.sh
Comment (fivetran-joemarkiewicz) [reply .buildkite/scripts/run_models.sh:22]: Yeah this was comment was made before I realized we took out all the incremental strategies. Let's remove the full refresh run as it isn't necessary

File: README.md
Comment (fivetran-joemarkiewicz) [reply README.md]: Following the discussion we are going to have these be disabled by default. Let's add this back in.

Comment (fivetran-joemarkiewicz) [reply README.md]: Also it looks like the formatting is a bit off here
image

File: README.md
Comment (fivetran-joemarkiewicz) [reply README.md]: If I recall, all of the below steps were only necessary due to the incremental logic (if you add a field the incremental models fail). Since we have no incremental logic, customers can add and subtract fields from the connector as much as they want. If this understanding is correct, we can remove the entire section below:
image

Comment (fivetran-joemarkiewicz) [reply README.md]: I also believe this section is now pointing to an out of date folder
image

File: README.md
Comment (fivetran-joemarkiewicz) [reply README.md:104]: That was a miss on my part, I meant this is not needed anymore. Yes, as the full create will always get the latest version of the source tables, the updating of the history models is not required.

File: dbt_project.yml
Comment (fivetran-joemarkiewicz) [reply dbt_project.yml:18]: Resolving this but making a note that we took a different approach

File: models/intermediate/workday_history/int_workday__employee_history.sql
Comment (fivetran-joemarkiewicz) [reply models/intermediate/workday_history/int_workday__employee_history.sql]: Great, thanks for the explanation this makes sense to me!

File: README.md
Comment (fivetran-joemarkiewicz) [reply README.md]: Also last topic on these - it looks like we have two "Step 4" sections. Could we correct the ordering or consolidate them. Whichever works best in your opinon.

File: README.md
Comment (fivetran-joemarkiewicz) [reply README.md:104]: Also one more note on the configurations of the history models. Can we include what the default value for `employee_history_start_date` is. (as of right now it is `2000-12-31`) and specify that these models will take the greater of this value and the min date in the source data.

Additionally, this was a date I arbitrarily picked. Do you feel there is a different date we should pick?

**Review by fivetran-joemarkiewicz — CHANGES_REQUESTED** [review]
@fivetran-avinash please see the notes above that have not been resolved. Those are my final remaining comments before this will be good to go! Let me know if you have any questions or if you would like to discuss anything in more detail. 

We are almost there!

File: README.md
Comment (fivetran-avinash) [reply README.md:104]: I'll moved it up to `2005-03-01`, the month when Workday was founded.

File: README.md
Comment (fivetran-avinash) [reply README.md]: Ahh I forgot to remove the border the first time. This is resolved now.

File: README.md
Comment (fivetran-avinash) [reply README.md]: Removed update section, updated folder, corrected steps ordering.

File: .buildkite/scripts/run_models.sh
Comment (fivetran-avinash) [reply .buildkite/scripts/run_models.sh:22]: Removed.

**Review by fivetran-avinash — COMMENTED** [review]
@fivetran-joemarkiewicz Ready for final review hopefully!

File: .buildkite/scripts/run_models.sh
Comment (fivetran-joemarkiewicz) [inline .buildkite/scripts/run_models.sh:23]: When running the test we need to turn on the variable. Otherwise the history tests will not run.

```suggestion
dbt test --vars '{employee_history_enabled: true}' --target "$db"
```

File: models/workday_history/workday__employee_daily_history.sql
Comment (fivetran-joemarkiewicz) [inline models/workday_history/workday__employee_daily_history.sql:30]: To match the comment above, let's just make this range truly be for one year. This is not necessary as it will not be used by a customer and only for compilation, but I want to ensure we don't get in trouble with causing unnescary confusion down the line.
```suggestion
    {% set start_date = dbt.dateadd("year", "-1", "current_date") %} -- One year in the past for first date
    {% set last_date = dbt.dateadd("day", "-1", "current_date") %} -- Yesterday as last date
```

File: models/workday_history/workday__worker_position_org_daily_history.sql
Comment (fivetran-joemarkiewicz) [inline models/workday_history/workday__worker_position_org_daily_history.sql:29]: Same suggestion from previous model for consistency and clarity for ourselves when we come back to this in the future.

```suggestion
{% else %} 
    {% set start_date = dbt.dateadd("year", "-1", "current_date") %} -- One year in the past for first date
    {% set last_date = dbt.dateadd("day", "-1", "current_date") %} -- Yesterday as last date
```

File: dbt_project.yml
Comment (fivetran-joemarkiewicz) [inline dbt_project.yml:13]: This is technically a breaking change as the old views are now going to be stale. I would call this out in the CHANGELOG as a breaking change and inform customers that the previous staging views will no longer be used.

**Review by fivetran-joemarkiewicz — APPROVED** [review]
@fivetran-avinash thanks so much for diligently working to get this update through the finish line and addressing all my review notes! 🎉 

I do have a few final suggestions and one comment to callout a breaking change that just occurred to me in the CHANGELOG. However, once those small updates are applied, this will be good to go! Approving now as I feel comfortable with you closing out these final changes and moving forward with release once those are applied and docs are regenerated!

File: dbt_project.yml
Comment (fivetran-joemarkiewicz) [reply dbt_project.yml:13]: Also, don't forget to add the link to the PR in the CHANGELOG for all these changes.

File: .buildkite/scripts/run_models.sh
Comment (fivetran-avinash) [reply .buildkite/scripts/run_models.sh:23]: Updated

File: README.md
Comment (fivetran-avinash) [reply README.md:102]: Updated

File: models/workday_history/workday__employee_daily_history.sql
Comment (fivetran-avinash) [reply models/workday_history/workday__employee_daily_history.sql:30]: Committed.

File: models/workday_history/workday__worker_position_org_daily_history.sql
Comment (fivetran-avinash) [reply models/workday_history/workday__worker_position_org_daily_history.sql:29]: Committed.

**Review by fivetran-avinash — COMMENTED** [review]
@fivetran-joemarkiewicz Final changes addressed!

**Pull request merged**
